From 70827c25b0f6d1d67c2db13d8323f1a18fd79446 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Mon, 4 Dec 2023 15:53:06 +0100
Subject: [PATCH 0815/1141] media: dcmipp: pixelproc: use
 pm_runtime_get_if_in_use in s/g_ctrl

Ensure that the registers can be accessed by using pm_runtime
around registers accesses in s/g_ctrl

Change-Id: I053de77d57adf39710aceef1c8e53ca7dfa0eb39
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/344624
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-pixelproc.c       | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
index eb50de90125c..ff50150f72f8 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
@@ -13,6 +13,7 @@
 #include <linux/module.h>
 #include <linux/mod_devicetable.h>
 #include <linux/platform_device.h>
+#include <linux/pm_runtime.h>
 #include <linux/vmalloc.h>
 #include <linux/v4l2-mediabus.h>
 #include <media/v4l2-ctrls.h>
@@ -203,6 +204,7 @@ struct dcmipp_pixelproc_device {
 	struct dcmipp_ent_device ved;
 	struct v4l2_subdev sd;
 	struct device *dev;
+	struct device *cdev;
 	struct v4l2_mbus_framefmt sink_fmt;
 	struct v4l2_mbus_framefmt src_fmt;
 	bool streaming;
@@ -253,12 +255,10 @@ static int dcmipp_pixelproc_s_ctrl(struct v4l2_ctrl *ctrl)
 			     struct dcmipp_pixelproc_device, ctrls);
 	int ret = 0;
 
-	mutex_lock(&pixelproc->lock);
+	if (pm_runtime_get_if_in_use(pixelproc->cdev) == 0)
+		return 0;
 
-	if (pixelproc->streaming) {
-		ret = -EBUSY;
-		goto out;
-	}
+	mutex_lock(&pixelproc->lock);
 
 	switch (ctrl->id) {
 	case V4L2_CID_PIXELPROC_GAMMA_CORRECTION:
@@ -267,8 +267,9 @@ static int dcmipp_pixelproc_s_ctrl(struct v4l2_ctrl *ctrl)
 		break;
 	}
 
-out:
 	mutex_unlock(&pixelproc->lock);
+	pm_runtime_put(pixelproc->cdev);
+
 	return ret;
 };
 
@@ -824,6 +825,7 @@ dcmipp_pixelproc_comp_bind(struct device *comp, struct device *master,
 
 	pixelproc->regs = bind_data->regs;
 	pixelproc->dev = comp;
+	pixelproc->cdev = master;
 
 	/* Initialize the lock */
 	mutex_init(&pixelproc->lock);
-- 
2.39.2

